File: AgqrPlayer4Tv/AgqrPlayer4Tv/Components/Fragments/PlaybackExoPlayerFragment.cs
using System;
using System.Reactive;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using AgqrPlayer4Tv.Components.Player;
using AgqrPlayer4Tv.Model.Platform;
using Android.OS;
using Android.Util;
using Android.Views;
using Com.Google.Android.Exoplayer;
using Reactive.Bindings;
using Reactive.Bindings.Extensions;

namespace AgqrPlayer4Tv.Components.Fragments
{
    public class PlaybackExoPlayerFragment : PlaybackPlayerFragmentBase, VideoPlayer.IListener
    {
        private const string Tag = "PlaybackExoPlayerFragment";

        private SurfaceView _surfaceView;
        private Subject<Tuple<bool, int>> _stateChanged = new Subject<Tuple<bool, int>>();
        private ReactiveProperty<VideoPlayer> _videoPlayer = new ReactiveProperty<VideoPlayer>();

        public IReadOnlyReactiveProperty<VideoPlayer> VideoPlayer { get { return this._videoPlayer; } }

        public PlaybackExoPlayerFragment()
        {
            this.IsPlaying = this.VideoPlayer
                    .SelectMany(_ => _stateChanged.Select(x => Unit.Default))
                    .Where(x => this.VideoPlayer.Value != null)
                    .Select(x => this.VideoPlayer.Value.PlaybackState == ExoPlayer.StateReady && this.VideoPlayer.Value.PlayWhenReady)
                    .ToReadOnlyReactiveProperty(false)
                    .AddTo(this.LifetimeDisposable);
        }

        public override View OnCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState)
        {
            var view = inflater.Inflate(Resource.Layout.PlaybackExoPlayer, container, false);

            this._surfaceView = (SurfaceView)view.FindViewById(Resource.Id.SurfaceView);
            return view;
        }

        public override void Play()
        {
            Log.Debug(Tag, $"Thread: {System.Threading.Thread.CurrentThread.ManagedThreadId}");
            if (this._surfaceView == null) return;

            this.Stop();

            var useRtmp = ApplicationMain.ServiceLocator.GetInstance<ApplicationPreference>().StreamingType.Value == StreamingType.Rtmp;
            var url = useRtmp
                ? UrlRtmp
                : UrlHls;
            var rendererBuilder = useRtmp
                ? (VideoPlayer.IRendererBuilder)new ExtractorRendererBuilder(this._surfaceView.Context, "App/1.0", Android.Net.Uri.Parse(url))
                : (VideoPlayer.IRendererBuilder)new HlsRendererBuilder(this._surfaceView.Context, UserAgent, url);
            Android.Util.Log.Debug(Tag, "Play: " + url);

            this._videoPlayer.Value = new VideoPlayer(rendererBuilder);
            this._videoPlayer.Value.Surface = this._surfaceView.Holder.Surface;
            this._videoPlayer.Value.AddListener(this);
            this._videoPlayer.Value.Prepare();

            this._videoPlayer.Value.PlayWhenReady = true;
        }

        public override void Stop()
        {
            if (this._videoPlayer.Value == null) return;

            this._videoPlayer.Value.PlayWhenReady = false;
            this._videoPlayer.Value.BlockingClearSurface();
            this._videoPlayer.Value.Release();
            this._videoPlayer.Value = null;
        }

        #region VideoPlayer.IListener Implementation

        void VideoPlayer.IListener.OnStateChanged(bool playWhenReady, int playbackState)
        {
            this._stateChanged.OnNext(Tuple.Create(playWhenReady, playbackState));
        }

        void VideoPlayer.IListener.OnError(Java.Lang.Exception e)
        {
            this.ReportError();
        }

        void VideoPlayer.IListener.OnVideoSizeChanged(int width, int height, int unappliedRotationDegrees, float pixelWidthHeightRatio)
        {
        }
        #endregion
    }
}

File: AgqrPlayer4Tv/app/src/main/java/org/misuzilla/agqrplayer4tv/component/fragment/PlaybackExoPlayerFragment.kt
package org.misuzilla.agqrplayer4tv.component.fragment

import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import org.misuzilla.agqrplayer4tv.R
import org.misuzilla.agqrplayer4tv.infrastracture.exoplayer.FlvExtractor2
import org.misuzilla.agqrplayer4tv.infrastracture.exoplayer.RtmpDataSource
import com.google.android.exoplayer2.*
import com.google.android.exoplayer2.ExoPlayer
import com.google.android.exoplayer2.source.ProgressiveMediaSource
import com.google.android.exoplayer2.source.TrackGroupArray
import com.google.android.exoplayer2.source.hls.HlsMediaSource
import com.google.android.exoplayer2.trackselection.AdaptiveTrackSelection
import com.google.android.exoplayer2.trackselection.DefaultTrackSelector
import com.google.android.exoplayer2.trackselection.TrackSelectionArray
import com.google.android.exoplayer2.ui.PlayerView
import com.google.android.exoplayer2.upstream.DefaultDataSourceFactory
import org.misuzilla.agqrplayer4tv.model.preference.ApplicationPreference
import org.misuzilla.agqrplayer4tv.model.preference.StreamingType

class PlaybackExoPlayerFragment : PlaybackPlayerFragmentBase(), Player.EventListener {
    private lateinit var exoPlayerView: PlayerView

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.playback_exoplayer, container, false)

        exoPlayerView = view.findViewById(R.id.surface_view)

        return view
    }

    override fun play() {
        val trackSelector = DefaultTrackSelector(requireContext(), AdaptiveTrackSelection.Factory())
        val loadControl = DefaultLoadControl()
        val exoPlayer = SimpleExoPlayer.Builder(requireContext())
            .setTrackSelector(trackSelector)
            .setLoadControl(loadControl)
            .build()
        val mediaSource = when (ApplicationPreference.getStreamingType().value) {
            StreamingType.HLS -> HlsMediaSource.Factory(DefaultDataSourceFactory(this.context, USER_AGENT)).createMediaSource(Uri.parse(URL_HLS))
            StreamingType.RTMP -> ProgressiveMediaSource.Factory({ RtmpDataSource() }, { arrayOf(FlvExtractor2()) }).createMediaSource(Uri.parse(URL_RTMP))
            else -> throw NotImplementedError()
        }

        exoPlayerView.apply {
            player = exoPlayer
            useController = false
        }

        exoPlayer.apply {
            addListener(this@PlaybackExoPlayerFragment)
            prepare(mediaSource)
            playWhenReady = true
        }
    }

    override fun stop() {
        exoPlayerView.player?.let {
            it.stop()
            it.release()
        }
        exoPlayerView.player = null
    }

    // ExoPlayer.EventListener
    override fun onPlayerError(error: ExoPlaybackException) {
        reportError()
    }

    override fun onLoadingChanged(isLoading: Boolean) {
    }

    override fun onPlayerStateChanged(playWhenReady: Boolean, playbackState: Int) {
        val player = exoPlayerView.player;
        if (player != null) {
            viewModel.setIsPlaying(player.playbackState == ExoPlayer.STATE_READY && player.playWhenReady)
        }
    }

    override fun onPlaybackParametersChanged(playbackParameters: PlaybackParameters) {
    }

    override fun onSeekProcessed() {
    }

    override fun onTracksChanged(trackGroups: TrackGroupArray, trackSelections: TrackSelectionArray) {
    }

    override fun onPositionDiscontinuity(reason: Int) {
    }

    override fun onRepeatModeChanged(repeatMode: Int) {
    }

    override fun onShuffleModeEnabledChanged(shuffleModeEnabled: Boolean) {
    }

    override fun onTimelineChanged(timeline: Timeline, reason: Int) {
    }
}